rfcs/0001-messages.md

```diff
@@ -0,0 +1,423 @@
+
+```rust
+pub trait Validate {
+    fn validate(&self) -> Result<&Self, MsgErr>;
+}
+```
+
+And we implement where needed. For example, in the header, sender and receiver can't be the same:
+
+```rust
+impl Validate for Header {
+    fn validate(&self) -> Result<&Self, MsgErr> {
+        if self.sender.0 == self.receiver.0 {
+            return Err(MsgErr::SameSenderAndReceiver);
+        }
+        Ok(self)
+    }
+}
+```
+
+This will require to have validation error messages as:
+
+```rust
+use thiserror::Error;
+
+#[derive(Clone, Error, Debug)]
+pub enum MsgErr {
+    #[error("sender and receiver are the same")]
+    SameSenderAndReceiver,
+}
+```
+
+Then to create a valid `Header` in the sender side we call:
+
+```rust
+let header = Validate::validate(&Header {
+    ..
+}).expect("a valid header");
+```
+
+The receiver side will validate the header using the same method. Instead of panicking the error can be ignored to don't crash and keep waiting for other (potentially valid) messages.
+
+```rust
+if let Ok(header) = msg.header.validate() {
+    ..
+}
+```
+
+### Rules
+
+The following rules must be implemented:
+
+#### Header
+
+- `version` must be a supported version.
+- `sender` and `receiver` can't be the same.
+- The `ParticipantId` variants of `sender` and `receiver` must match the message type.
+
+#### Payloads
+
+- Each jubjub type must be validated during deserialization.
+- `share_commitments`:
+  - Length must be less than or equal to `MAX_SIGNER_PARTICIPANT_ID`.
+  - Lenght must be at least `MIN_SIGNERS` (`2` signers).
+- `signing_commitments`:
+    - Length must be less than or equal to `MAX_SIGNER_PARTICIPANT_ID`.
+    - Signing packages that contain duplicate `ParticipantId`s are invalid. This is implicit in the use of `HashMap`.
+    - Length must be at least `MIN_THRESHOLD` (`2` required signers).
+- `message`: signed messages have a protocol-specific length limit. For Zcash, that limit is the maximum network protocol message length: `2^21` bytes (2 MB).
+
+## Serialization/Deserialization
+
+Each message struct needs to serialize to bytes representation before it is sent through the wire and must deserialize to the same struct (round trip) on the receiver side. We use `serde` and macro derivations (`Serialize` and `Deserialize`) to automatically implement where possible.
+
+This will require deriving serde in several types defined in `frost.rs`. 
+Manual implementation of serialization/deserialization will be located at a new mod `src/frost/serialize.rs`.
+
+### Byte order
+
+Each byte chunk specified below is in little-endian order unless is specified otherwise.
+
+Multi-byte integers **must not** be used for serialization, because they have different byte orders on different platforms.
+
+### Header
+
+The `Header` part of the message is 4 bytes total:
+
+Bytes | Field name | Data type
+------|------------|-----------
+1     | msg_type   | u8
+1     | version    | u8
+1     | sender     | u64
+1     | receiver   | u64
```

Comment:
Two urgent changes:
1. The version field must come first, so the parser knows how to interpret the rest of the message
2. Update the `ParticipantId` sizes

Another change we should open a ticket for:
3. Consider aligning all fields to 8 bytes using padding (or by expanding other fields)
```suggestion
The `Header` part of the message is 18 bytes total:

Bytes | Field name | Data type
------|------------|-----------
1     | version    | u8
1     | msg_type   | u8
8     | sender     | u64
8     | receiver   | u64
```